_posts/2017-08-28-the-rx-way.md

```diff
@@ -0,0 +1,123 @@
+
+Now that is lot of work and I am not even considering edge cases and error handling for sake of simplicity. Think of the case when API returns a large list of objects and you can't afford to process that on main thread.
+
+Well, Rx does all this heavy lifting for you. Lets see how.
+
+### The Rx Way
+
+Reactive programming is totally a different approach towards writing code. Everything in reactive world is considered a stream. These streams are immutable and there are plethora of operators available which operate on these streams. Real power of reactive programming lies in these tried and tested operators.
+
+Rx makes task of switching between the thread as easy as having an apple pie. Our above said problem statement can be accomplished by mere 10 lines of code with RxJava. Here is a sample from production code
+
+{% highlight java %}
+APIProvider.getInstance().getService().getMyObjects(userName)
+  .subscribeOn(Schedules.io())
+  .map(MyObject::toList)
+  .flatMap(rx.Observable::from)
+  .filter(myObject->myObject.getName().equals(Constants.SOME_NAME))
+  // adding a 10ms delay in subsequent api calls
+  .flatMap(myObject -> Observable.timer(100, TimeUnit.MILLISECONDS).map(y -> myObject))
+  .doOnNext(rule -> {
+      // deleting my object one by one
+      APIProvider.getInstance()
+        .getService().deleteMyObject(userName, rule.getId())
+        .subscribe();
+  }).subscribe();
+{% endhighlight %}
+
+lets understand this code line by line
+
+```
+APIProvider.getInstance().getService().getMyObjects(userName)
+```
+Its `Retrofit` way of doing an API call. This API returns a map of `MyObjectId` to `MyObject`. An example JSON looks like this
+{% highlight json %}
+{
+  "1": {
+    "name": "object1"
+  },
+  "2": {
+    "name": "object2"
+  }
+}
+{% endhighlight %}
+
+```
+.subscribeOn(Schedules.io())
+```
+`subscribeOn` is an Rx operator which directs RxJava to start processing on scheduler thread (a background thread). Just this line lets you switch your processing to a background thread. How cool is that!
+
+```
+.map(MyObject::toList)
+```
+`Map` is transformation operator which allows us to convert things. Here we are converting `Map` to `List`. `toList` is a static method written in `MyObject` class. This method converts a `Map<ObjectId, MyObject>` to `List<MyObject>`. List version has object id's in object itself. `MyObject::toList` is Java 8 way of calling this method, Its named as [Method Reference](https://docs.oracle.com/javase/tutorial/java/javaOO/methodreferences.html).
+
+```
+.flatMap(rx.Observable::from)
+```
+
+This statement allows us to receive each `MyObject` as separate event on observable stream. Before this we were having whole List or Map as single event on observable stream. To perform filtering and API calls on each of these separately we need to have them as independent events on stream.
```

Comment:
Maybe you can explain here how stream of stream is working and how we are again combining them?

Reply:
My primary objective of this post was to demonstrate reactive programming abilities. I do not want to dive much into how Rx works but what its capable of doing. This is in order to keep things simple for reader. Do you think this explanation is absolutely needed for fulfilling primary objective?

Reply:
I think this should be enough to spark the curiosity of the casual reader, and that is what we want to do. An in depth discussion of streams can warrant its own post.